object-common-multipart.go

```diff
@@ -192,3 +220,305 @@ func abortMultipartUploadCommon(storage StorageAPI, bucket, object, uploadID str
+			if fileInfo.Mode.IsDir() {
+				// If a directory is found, doesn't return anything.
+				return nil, nil
+			}
+			fileName := path.Base(fileInfo.Name)
+			if !strings.Contains(fileName, ".") {
+				// Skip the entry if it is of the pattern bucket/object/uploadID.partNum.md5sum
+				// and retain entries of the pattern bucket/object/uploadID
+				entries = append(entries, fileInfo)
+			}
+		}
+		if eof {
+			break
+		}
+	}
+	return entries, nil
+}
+
+// listMetaBucketMultipartFiles - list all files at a given prefix inside minioMetaBucket.
+func listMetaBucketMultipartFiles(storage StorageAPI, prefixPath string, markerPath string, recursive bool, maxKeys int) (allFileInfos []FileInfo, eof bool, err error) {
+	// newMaxKeys tracks the size of entries which are going to be
+	// returned back.
+	var newMaxKeys int
+
+	// Following loop gathers and filters out special files inside
+	// minio meta volume.
+	for {
+		var fileInfos []FileInfo
+		// List files up to maxKeys-newMaxKeys, since we are skipping
+		// entries for special files.
+		fileInfos, eof, err = storage.ListFiles(minioMetaBucket, prefixPath, markerPath, recursive, maxKeys-newMaxKeys)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"prefixPath": prefixPath,
+				"markerPath": markerPath,
+				"recursive":  recursive,
+				"maxKeys":    maxKeys,
+			}).Errorf("%s", err)
+			return nil, true, err
+		}
+		// Loop through and validate individual file.
+		for _, fi := range fileInfos {
+			var entries []FileInfo
+			if fi.Mode.IsDir() {
+				// List all the entries if fi.Name is a leaf directory, if
+				// fi.Name is not a leaf directory then the resulting
+				// entries are empty.
+				entries, err = listLeafEntries(storage, fi.Name)
+				if err != nil {
+					log.WithFields(logrus.Fields{
+						"prefixPath": fi.Name,
+					}).Errorf("%s", err)
+					return nil, false, err
+				}
+			}
+			// Set markerPath for next batch of listing.
+			markerPath = fi.Name
+			if len(entries) > 0 {
+				// We reach here for non-recursive case and a leaf entry.
+				for _, entry := range entries {
+					allFileInfos = append(allFileInfos, entry)
+					newMaxKeys++
+					// If we have reached the maxKeys, it means we have listed
+					// everything that was requested. Return right here.
+					if newMaxKeys == maxKeys {
+						// Return values:
+						// allFileInfos : "maxKeys" number of entries.
+						// eof : eof returned by fs.storage.ListFiles()
+						// error : nil
+						return
+					}
+				}
+			} else {
+				// We reach here for a non-recursive case non-leaf entry
+				// OR recursive case with fi.Name matching pattern bucket/object/uploadID[.partNum.md5sum]
+				if !fi.Mode.IsDir() { // Do not skip non-recursive case directory entries.
+					// Skip files matching pattern bucket/object/uploadID.partNum.md5sum
+					// and retain files matching pattern bucket/object/uploadID
+					specialFile := path.Base(fi.Name)
+					if strings.Contains(specialFile, ".") {
```

Comment:
This is dangerous assumption. Anyways fix is to use the new special filename based detection.